proposals/debug-id.md

```diff
@@ -0,0 +1,237 @@
+the Debug ID in hexadecimal characters, preferably in the canonical UUID
+format:
+
+```json
+{
+  "version": 3,
+  "file": "app.min.js",
+  "debugId": "85314830-023f-4cf1-a267-535f4e37bb17",
+  "sources": [...],
+  "sourcesContent": [...],
+  "mappings": "..."
+}
+```
+
+### Debug IDs in JavaScript Artifacts
+
+Transpiled JavaScript files containing a Debug ID must embed the ID near the end
+of the source, ideally on the last line, in the format `//# debugId=<DEBUG_ID>`:
+
+```javascript
+//# debugId=85314830-023f-4cf1-a267-535f4e37bb17
+```
+
+If the special `//# sourceMappingURL=` comment already exists in the file, it is
+recommended to place the `debugId` comment in the line above to maintain
+compatibility with existing tools.  Because the last line already has meaning in
+the existing specification for the `sourceMappingURL` comment, tools are
+required to examine the last 5 lines to discover the Debug ID.
+
+## JavaScript API for Debug ID Resolution
+
+Today `error.stack` in most runtimes only returns the URLs of the files referenced
+by the stack trace.  For Debug IDs to be useful, a solution would need to be added
+to enable mapping of JavaScript file URLs to Debug IDs.
+
+The strawman proposal is to add the Debug ID in two locations:
+
+* `import.meta.debugId`: a new property that should return the debug ID as UUID
+  of the current module if has one
+* `System.getDebugIdForUrl(url)` looks up the debug ID for a given script file by
+  URL that has already been loaded by the browser in the current context.
+
+## Appendix A: Self-Description of Files
+
+Unfortunately, neither transpiled JavaScript files nor source maps can be easily
+identified without employing heuristics. Unlike formats like ELF binaries, they
+lack a distinctive header for identification purposes. When batch processing
+files, the ability to differentiate between various files is invaluable, but
+this capability is not fully realized in the context of source maps or
+transpiled JavaScript files. Although solving this issue is beyond the scope of
+this document, addressing it would significantly aid in distinguishing different
+files without relying on intricate heuristics.
+
+Nevertheless, we recommend that tools utilize the following heuristics to
+determine self-identifying JavaScript files and source maps:
+
+* a JSON file containing a toplevel object with the keys `mapping`, `version`,
+  `debugId` and `sourcesContent` should be considered to be a self-identifying
+  source map.
+* a UTF-8 encoded text file matching the regular expression
+  `(?m)^//# debugId=([a-fA-F0-9-]{12,})$` should be considered a transpiled
+  JavaScript file.
+
+## Appendix B: Symbol Server Support
+
+With debug IDs it becomes possible to resolve source maps and minified JavaScript
+files from the server.  That way a tool such as a browser or a crash reporter could
+be pointed to a S3, GCS bucket or an HTTP server that can serve up source maps and
+build artifacts keyed by debug id.
```

Comment:
I believe we should also plan for browsers and local development servers.
Using something like Import Maps could solve this, and point the browser to the local development server to resolve the corresponding source map.

Related to my note above, this could potentially replace the various different ways that bundlers / development servers can offer source maps today. I believe offering a local symbol server would also benefit performance a bit, as the tools wouldn’t have to embed base-64 encoded sourcemaps into the development assets, but sourcemaps would only be needed to be serialized on demand.